### PR TITLE
Password setup tool for built-in users

### LINKED ISSUE #1576: Replace `admin:admin` default credentials
**Is your feature request related to a problem? Please describe.**
The default `admin:admin` credential is too weak and a sure-fire way to get hacked. Though the documentation recommends users to change the default passwords, it can be easily missed. Such weak defaults can be very easily exploited and lead to massive data leaks (remember [Equifax attack](https://www.forbes.com/sites/kateoflahertyuk/2019/10/20/equifax-lawsuit-reveals-terrible-security-practices-at-time-of-2017-breach/?sh=446983d93d38)?). 

The second gap is, in the existing distribution, there's no default password policy to ensure stronger passwords. It needs to be configured explicitly by users. That is why such weak credentials are able to slip through the cracks.

**Describe the solution you'd like**
Remove the default passwords provided in `internal_users.yml`. Instead, have the `admin` user setup passwords for internal users once the OpenSearch process is up. These passwords can be set up post-installation using a REST API and `admin` certificates (for authentication). Setting password for `admin` user would look something like -
```
PATCH _opendistro/_security/api/internalusers/admin --key admin-key.pem --cert admin.pem --cacert root-ca.pem -d'
{
    "password": "quCn52_8\*N8"
}
'
```
To make this further easy, OpenSearch distribution could provide a tool such as an OpenSearch shell client to make it easy for admin user to configure all users. This shell client could also help with autogenerating strong passwords and can be further extended to support for more administrative functions in future. 
 
In addition to this, we should also add a default password policy to only allow passwords that match a certain strength criteria. Obv, these policies are overridable depending on the user/enterprise needs. Having a strong default policy will protect against insecure password configurations.

**Describe alternatives you've considered**

**Additional context**
This is related to https://github.com/opensearch-project/OpenSearch/issues/1618

**Task breakdown**
- [x] Remove default passwords for inbuilt users in internal_users.yml, deprecate now and remove password field from yml in 2.0
- [x] Configure strong default password policy and also make it overridable via yml.
- [x] Write tool for password setup (to be used by admin).
- [x] Add auto-password-generation logic to the tool to ease user experience during setup.
- [ ] #1787

### PR DESCRIPTION
**Pull request by setiah**
### Description
This PR primarily provides new password setup tool for setting up passwords for inbuilt security users (admin, kibanaserver, etc.) and also provides a new unified interface for extending tooling in opensearch-security. 

> Details of all the changes
1. Provides a new unified interface to simplify and extend tooling in opensearch-security.
2. Adds setup-password functionality to allow users to setup secure passwords for built-in users, instead of relying on weak defaults such as admin:admin. The functionality is provided in both manual and auto mode for auto-password generation.
3. Adds default secure password policy to opensearch.yml

> What is the old behavior before changes and new behavior after changes?

In the old behavior, users had to update the internal_users.yml file on all nodes with the hash of the password to change the passwords from admin/admin. While that still works, with the new tool, they can simply run the tool to setup secure passwords after installation. The tool runs as `opensearch-admin setup-passwords [--auto]` or to setup policy compliant passwords for inbuilt users (both manually and automatically). It can be extended to support other functions such as `setup-certificates`, `reload-configs`, through its simplified extensible interface. 

### Issues Resolved
#1576 

Is this a backport? If so, please add backport PR # and/or commits #
No
### Testing
Unit testing, Manual testing

### Check List
- [x] New functionality includes testing
- [x] Commits are signed per the DCO using --signoff
- [ ] New functionality has been documented

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by setiah** [comment]
@peternied can you help with reviewing this PR? 
cc: @ryanbogan

**Comment by setiah** [comment]
@peternied can you help with reviewing this PR? 
cc: @ryanbogan

**Comment by CEHENKLE** [comment]
@opensearch-project/security Hey folks -- can we get some code review luv over here?  

Thanks :) 

/C

File: src/test/java/org/opensearch/security/SlowIntegrationTests.java [hunk 1]
```diff
@@ -30,6 +30,7 @@
 
 package org.opensearch.security;
 
+
```
Comment (cliu123) [inline src/test/java/org/opensearch/security/SlowIntegrationTests.java:33]: Please remove this empty line if not necessary.

**Review by peternied — CHANGES_REQUESTED** [review]
Thanks for the pull request.  Making sure that new clusters have strong security posture is an important space for us.

This pull request changes the responsibility of the security plugin by adding a new OpenSearch-Admin script.  This is a public interface in the security repository that over time will likely need to be deprecated based on de-emphasizing security as a plugin.  We should avoid this, I would recommend baking these changes into a cluster initialization process.

Accepting custom passwords creates the opportunity for bad password practices.  OpenSearch would have a better posture if we had a standard password generation of say 20+ characters of entropy.  This could be done with modifications to the existing create user API or as part of the initialization process to standard out.

Please reduce the scope of this PR to an automatic (strong) password generation during the initialization process, and extract the admin tool/script creation into a separate draft PR so we can have that larger discussion over there, and weigh the pros and cons of adding a tool that will need to be deprecated vs postponing that effort until we have a broader plan for admin tools moving forward.

**Comment by setiah** [comment]
@peternied 

There is a separate [issue](https://github.com/opensearch-project/security/issues/1787) tracking auto-password setup during cluster bootstrap, which leverages the PasswordSetup logic added by this tool to autogenerate passwords. But passwords are not just setup during the first time setup. They also need to be rotated many times during the lifecycle of a cluster based on security needs of an organization. This tool provides an easy way to for cluster administrator to set the passwords first time (until we have [this](https://github.com/opensearch-project/security/issues/1787)) and rotate/update them later. Hence it would be helpful.

>This pull request changes the responsibility of the security plugin by adding a new OpenSearch-Admin script. This is a public interface in the security repository that over time will likely need to be deprecated based on de-emphasizing security as a plugin. 

The public interface is for Security plugin tooling and is not needed in other plugins. All tools in Security plugin (hasher, securityadmin, install_demo_configurations, etc.) can extend this interface to inherit some common functions, and just build their business logic. This will enable java-fying the security plugin tools so they are unit testable like code, unlike some shell tools in security plugin repo today that need to be validated manually. Happy to consider if you have any other suggestions for the tool name.

>Accepting custom passwords creates the opportunity for bad password practices. This could be done with modifications to the existing create user API or as part of the initialization process to standard out.

I think you missed this change mentioned in the PR description - "Adds default secure password policy to opensearch.yml". The custom passwords are validated against this default secure password policy that is being adding with this change. If the password is not policy compliant, it gets rejected. 
Lack of a default secure password policy is a gap today which allows users to create weak passwords 

I'd request you to take a look again.

**Comment by peternied** [comment]
> This tool provides an easy way to for cluster administrator to set the passwords first time (until we have https://github.com/opensearch-project/security/issues/1787) and rotate/update them later. Hence it would be helpful.

The security plugin already has an API, there is no extra burden for documentation, maintenance or release management.  A new endpoint such as  `PUT user/{user}/rotateCredentials`, or improving existing endpoints for this scenario seems very reasonable.  Exposing this functionality with an API expands the utility so admins could make API calls directly, use the Dashboard UX, or use another plugin to go through the same workflow.  If this is built as a tool its utility is limited, lets build this functionality so it can be reused.  Is there a reason to limit the functionality?

> I think you missed this change mentioned in the PR description - "Adds default secure password policy to opensearch.yml". 

This change was done in the `tools/install_demo_configuration.sh` file, outside of the recommended configuration.  There would be much more benefit to make this change as part of the security default configuration.  What do you think about separating password policy into another pull request?

**Comment by setiah** [comment]
> The security plugin already has an API, there is no extra burden for documentation, maintenance or release management. A new endpoint such as `PUT user/{user}/rotateCredentials`, or improving existing endpoints for this scenario seems very reasonable. Exposing this functionality with an API expands the utility so admins could make API calls directly, use the Dashboard UX, or use another plugin to go through the same workflow. If this is built as a tool its utility is limited, lets build this functionality so it can be reused. Is there a reason to limit the functionality?

The functionality already exists today with a PATCH API. The tool just builds on the capability of existing PATCH API to simplify password setup process for all internal users. I do think there is value in having a tool based mechanism over and above REST API as it makes life easier for users. 


> This change was done in the `tools/install_demo_configuration.sh` file, outside of the recommended configuration. There would be much more benefit to make this change as part of the security default configuration. What do you think about separating password policy into another pull request?

I'm okay with this suggestion.

**Pull request closed without merging**